Framework/PythonInterface/test/python/mantid/utils/dgsTest.py

```diff
@@ -0,0 +1,462 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+import unittest
+from numpy import array, nan, where, vstack
+from numpy.testing import assert_almost_equal, assert_array_equal
+from mantid.utils.dgs import qangle, ErrorCodes
+from mantid.geometry import OrientedLattice
+
+class DgsUtilsTest(unittest.TestCase):
+
+    def run_test(self, ol, expected_Qx, expected_Qy, expected_Qz, expected_in_plane_Q_angle, expected_out_plane_Q_angle,
+                 expected_in_plane_kf_angle, expected_out_plane_kf_angle, expected_omega, expected_error_code, **pars):
+
+        result = qangle(lattice=ol, **pars)
+        good_results = where(result.error_code == ErrorCodes.CORRECT)
+        assert_almost_equal(expected_Qx[good_results], result.Q_lab_x[good_results],2)
+        assert_almost_equal(expected_Qy[good_results], result.Q_lab_y[good_results],2)
+        assert_almost_equal(expected_Qz[good_results], result.Q_lab_z[good_results],2)
+        assert_almost_equal(expected_in_plane_Q_angle[good_results], result.in_plane_Q_angle[good_results], 1)
+        assert_almost_equal(expected_out_plane_Q_angle[good_results], result.out_plane_Q_angle[good_results], 1)
+        assert_almost_equal(expected_in_plane_kf_angle[good_results], result.in_plane_kf_angle[good_results], 1)
+        assert_almost_equal(expected_out_plane_kf_angle[good_results], result.out_plane_kf_angle[good_results], 1)
+        assert_almost_equal(expected_omega[good_results], result.omega[good_results], 1)
+        assert_array_equal(expected_error_code, result.error_code)
+
+    def test_qangle_NoGC_NoOC(self):
+        # No geometry constraints, no omega constraints, all parameters are arrays
+        test_case1 = array([[10, 0, 1, 0, 0, 1, -0.72, 0, 0.12, 19, 0, -80.5, 0, -80.5, 0],
```

Comment:
For future developers / people writing the test could we leave a note on how / where this data comes from? If it's generated how would say someone else generate another similar dataset to add an additional test, since it's a unusually large "hand-typed" array

Reply:
Older (buggier) instrument scientist developed code. The IS created CSV files in Excel, but I'dve decided against adding 9 extra files to our tests.